File: compiler/bench/countingsort.c
// expected: 49950000
// Counting sort benchmark - sort 100K elements

#include <stdio.h>

#define SIZE 100000
#define RANGE 1000

int arr[SIZE];
int cnt[RANGE];
int out[SIZE];

void counting_sort(void) {
    // Clear counts
    for (int i = 0; i < RANGE; i++) cnt[i] = 0;

    // Count occurrences
    for (int i = 0; i < SIZE; i++) {
        cnt[arr[i]]++;
    }

    // Cumulative count (descending order)
    for (int i = RANGE - 2; i >= 0; i--) {
        cnt[i] += cnt[i + 1];
    }

    // Build output
    for (int i = 0; i < SIZE; i++) {
        out[--cnt[arr[i]]] = arr[i];
    }

    // Copy back
    for (int i = 0; i < SIZE; i++) {
        arr[i] = out[i];
    }
}

void init_arr(void) {
    for (int i = 0; i < SIZE; i++) {
        arr[i] = i % RANGE;
    }
}

long sum_arr(void) {
    long sum = 0;
    for (int i = 0; i < SIZE; i++) {
        sum += arr[i];
    }
    return sum;
}

int main(void) {
    init_arr();
    counting_sort();
    printf("%ld\n", sum_arr());
    return 0;
}

File: compiler/bench/countingsort.fs
\ expected: 49950000
\ Counting sort benchmark - sort 100K elements

100000 constant SIZE
1000 constant RANGE
create arr SIZE cells allot
create cnt RANGE cells allot
create out SIZE cells allot

: arr@ ( i -- n ) cells arr + @ ;
: arr! ( n i -- ) cells arr + ! ;
: cnt@ ( i -- n ) cells cnt + @ ;
: cnt! ( n i -- ) cells cnt + ! ;
: out@ ( i -- n ) cells out + @ ;
: out! ( n i -- ) cells out + ! ;

: counting-sort ( -- )
  \ Clear counts
  RANGE 0 do 0 i cnt! loop
  \ Count occurrences
  SIZE 0 do
    i arr@ dup cnt@ 1+ swap cnt!
  loop
  \ Cumulative count (descending order)
  RANGE 2 - 0 swap do
    i cnt@ i 1+ cnt@ + i cnt!
  -1 +loop
  \ Build output
  SIZE 0 do
    i arr@                \ val
    dup cnt@ 1-           \ val idx
    dup rot cnt!          \ idx ; update count
    i arr@ swap out!      \ store val at out[idx]
  loop
  \ Copy back
  SIZE 0 do i out@ i arr! loop ;

: init-arr ( -- )
  SIZE 0 do
    i RANGE mod i arr!
  loop ;

: sum-arr ( -- n )
  0 SIZE 0 do i arr@ + loop ;

: main
  init-arr
  counting-sort
  sum-arr . cr ;
